File: Java/15.3Sum.java

public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (nums == null || nums.length < 2){
            return res;
        }
        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i ++){
            if (i != 0 && nums[i] == nums[i - 1]){
                continue;
            }
            int start = i + 1;
            int end = nums.length - 1;

            while( start < end){
                int sum = nums[i] + nums[start] + nums[end];
                if ((sum) == 0){
                    List<Integer> tmp = new ArrayList<Integer>();
                    tmp.add(nums[i]);
                    tmp.add(nums[start]);
                    tmp.add(nums[end]);
                    res.add(tmp);
                    start ++;
                    end --;
                    //jump away from dumplicate nums
                    while( start < end  && nums[start] == nums[start - 1]){
                        start++;
                    }
                    while(end > start && nums[end] == nums[end + 1]){
                        end --;
                    }
                }                
                else if (sum > 0){
                    end --;
                }else{
                    start++;
                }
            }
        }
        return res;
    }
}

File: Cpp/15.3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if(nums.size() < 3){
            return {};
        }
        
        sort(nums.begin(), nums.end());
        
        set<vector<int>> resultSet;
        int length = nums.size();
        
        for(int i = 0; i < length-2; ++i){
            if(i > 0 && nums[i] == nums[i-1]){
                continue;
            }
            
            int left = i+1;
            int right = length-1;
            int target = 0 - nums[i];
            while(left < right){
                int leftNum = nums[left];
                int rightNum = nums[right];
                if(leftNum + rightNum > target){
                    while(right-1 == right){
                        right--;
                    }
                    right--;
                    
                }else if(leftNum + rightNum < target){
                    while(left+1 == left){
                        left++;
                    }
                    left++;
                }else{
                    resultSet.insert({nums[i], leftNum, rightNum});
                    while(right-1 == right){
                        right--;
                    }
                    right--;
                    while(left+1 == left){
                        left++;
                    }
                    left++;
                }
            }
        }
        
        vector<vector<int>> result(resultSet.begin(), resultSet.end());
        
        return result;
    }
};